### PR TITLE
move CRD behind TPR

### PR DESCRIPTION
**Pull request by deads2k**
Second attempt at https://github.com/kubernetes/kubernetes/pull/46297

@liggitt since @sttts is out can you take a look and hopefully find me a different, more sane way to do this? It's a little crazy, but I left a large comment explaining why I think its the only way.

### CODE REVIEW

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -105,8 +105,18 @@ func Run(runOptions *options.ServerRunOptions, stopCh <-chan struct{}) error {
 		return err
 	}
 
-	// kubeAPIServer is at the base for now.  This ensures that CustomResourceDefinitions trump TPRs
-	kubeAPIServer, err := CreateKubeAPIServer(kubeAPIServerConfig, genericapiserver.EmptyDelegate, sharedInformers)
+	// TPRs are enabled and not yet beta, since this these are the successor, they fall under the same enablement rule
+	// Subsequent API servers in between here and kube-apiserver will need to be gated.
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: subsequent reads weirdly now that this follows kube-apiserver

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 2]
```diff
@@ -98,6 +120,51 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+	name               string
+	goRestfulContainer *restful.Container
+	postGoRestfulMux   *mux.PathRecorderMux
+}
+
+func (d director) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	path := req.URL.Path
+
+	// we only need further inspection on paths under /apis
+	if !strings.HasPrefix(path, "/apis") {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: `/apis/`, right?

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 3]
```diff
@@ -98,6 +120,51 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+		return
+	}
+
+	// at this point we may or may not need to risk gorestful. We need to do it if there is a webservice root path
+	// other than /apis which is a prefix for our request.  Otherwise we skip it.
+	for _, ws := range d.goRestfulContainer.RegisteredWebServices() {
+		if ws.RootPath() == "/api" || ws.RootPath() == "/apis" || ws.RootPath() == "/" {
+			continue
+		}
+		if strings.HasPrefix(path, ws.RootPath()) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: are there webservices registered at both the group and version level? without trailing slashes (so they match requests with trailing slashes)? do we run into trouble if a CRD exists for group `/apis/extensions-with-suffix.acme.com` and the core apiserver gets a prefix hit on `/apis/extensions`?

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 4]
```diff
@@ -98,6 +120,51 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+	}
+
+	// if we didn't find a match, then we just skip gorestful altogether
+	glog.V(5).Infof("%v: %q satisfied postGoRestful", d.name, path)
+	d.postGoRestfulMux.ServeHTTP(w, req)
+}
+
+// getRelaventGoRestfulWebServicePaths uses an atomic value updated based on a timer to handle additions in a timely fashion
+// without requring r
+func (d director) getRelaventGoRestfulWebServicePaths() {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: unused? also, typo

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go [hunk 5]
```diff
@@ -130,7 +134,18 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 
 	crdClient, err := internalclientset.NewForConfig(s.GenericAPIServer.LoopbackClientConfig)
 	if err != nil {
-		return nil, err
+		// it's really bad that this is leaking here, but until we can fix the test (which I'm pretty sure isn't even testing what it wants to test),
```
Comment (liggitt) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go]: is this behavior new in this PR?

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (deads2k) [inline cmd/kube-apiserver/app/server.go]: ok

File: staging/src/k8s.io/apiserver/pkg/server/handler.go (hunk 2 above)
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: ok

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go (hunk 5 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go]: kind of.  Before when you set that it would fail differently.  Now its on a critical path.  That entire env var needs rethinking, but its bigger than any modification we'd do in 1.7.

File: staging/src/k8s.io/apiserver/pkg/server/handler.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: They register without slashes.  Updating to `if path == ws.RootPath() || strings.HasPrefix(path, ws.RootPath()+"/") `

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 6]
```diff
@@ -48,14 +49,28 @@ type APIServerHandler struct {
 	// PostGoRestfulMux is the final HTTP handler in the chain.
 	// It comes after all filters and the API handling
 	PostGoRestfulMux *mux.PathRecorderMux
+
+	// Director is here so that we can properly handle fall through and proxy cases.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: let me explain. no, there is too much. let me sum up.

explaining the data flow of the four handlers (and maybe tweaking names) would help enormously:
```
FullHandlerChain -> Director -> {GoRestfulContainer,PostGoRestfulMux} based on inspection of registered web services
GoRestfulContainer -> {web services, PostGoRestfulMux}
```

also, what each should be used for externally, and who is responsible for building each

**Comment by liggitt** [comment]
needs a test proving TPR wins when a CRD and TPR both claim the same group/version/resource

also would like a test around the prefix-matching issue (CRD for an `extensions.acme.com` API group should work)

**Comment by deads2k** [comment]
> needs a test proving TPR wins when a CRD and TPR both claim the same group/version/resource

An integration test won't work since kube integration tests don't start a kube-apiserver.  An e2e doesn't have the clients we need, though that may eventually be surmountable.  How about demonstrating that the prefix matching wins in a unit test and you trust the unit test and the clear gorestful registration order.

File: staging/src/k8s.io/apiserver/pkg/server/handler.go (hunk 3 above)
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: that's a lot of string allocations. (assuming all webservice rootpaths have no trailing slash) what about:

```go
if strings.HasPrefix(path, ws.RootPath()) {
  // ensure an exact match or a path boundary match
  if len(path) == len(ws.RootPath()) || path[len(ws.RootPath())] == '/' {
     ...
  }
}
```

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: I think it uglifies code in anticipatory (but unproven) performance problems, but I will relent to get the overall pull in.

**Comment by deads2k** [comment]
@enisoc if you need something to play with, I think this works as long as you don't create any TPRs.  Looks like we're hitting a gorestful bug that I'll code around for 1.7.

**Comment by deads2k** [comment]
Comments addressed.

File: hack/make-rules/test-cmd.sh [hunk 7]
```diff
@@ -77,7 +77,7 @@ function run_kube_controller_manager() {
 # Creates a node object with name 127.0.0.1. This is required because we do not
 # run kubelet.
 function create_node() {
-  kubectl create -f - -s "http://127.0.0.1:${API_PORT}" << __EOF__
+  kubectl create -f - -s "https://127.0.0.1:${SECURE_API_PORT}" --token=admin/system:masters --insecure-skip-tls-verify=true << __EOF__
```
Comment (liggitt) [inline hack/make-rules/test-cmd.sh]: why was this needed?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 8]
```diff
@@ -251,7 +251,7 @@ func DefaultOpenAPIConfig(getDefinitions openapicommon.GetOpenAPIDefinitions, sc
 // WebServices set.
 func DefaultSwaggerConfig() *swagger.Config {
 	return &swagger.Config{
-		ApiPath:         "/swaggerapi/",
+		ApiPath:         "/swaggerapi",
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: is this fixing a double-slash issue, or changing consumption of subpaths, or something else?

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 9]
```diff
@@ -37,27 +38,45 @@ import (
 )
 
 // APIServerHandlers holds the different http.Handlers used by the API server.
-// This includes the full handler chain, the gorestful handler (used for the API) which falls through to the postGoRestful handler
+// This includes the full handler chain, the director (which chooses between gorestful and postGoRestful,
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: s/postGoRestful/nonGoRestful/

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 8 above)
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: > is this fixing a double-slash issue, or changing consumption of subpaths, or something else?

slash check for the non-allocating test.

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 10]
```diff
@@ -92,12 +114,57 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+	path := req.URL.Path
+
+	// if we are exactly /apis or /apis/, pass then we need special handling in loop.
+	// normally these are passed to the nonGoRestfulMux, but if discovery is enabled, it will go directly.
+	// We can't rely on a prefix match since /apis matches everything (see the big comment on Director above)
+	isDiscoveryRequest := path == "/apis" || path == "/apis/"
+
+	// check to see if our webservices want to claim this path
+	for _, ws := range d.goRestfulContainer.RegisteredWebServices() {
+		if ws.RootPath() == "/apis" {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: making it clear the two blocks are mutually exclusive would make it a little easier to read the loop:
```
switch {
case ws.RootPath() == "/apis":
  ...
case strings.HasPrefix(path, ws.RootPath()):
  ...
}
```

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 11]
```diff
@@ -92,12 +114,57 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+	nonGoRestfulMux    *mux.PathRecorderMux
+}
+
+func (d director) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	path := req.URL.Path
+
+	// if we are exactly /apis or /apis/, pass then we need special handling in loop.
+	// normally these are passed to the nonGoRestfulMux, but if discovery is enabled, it will go directly.
+	// We can't rely on a prefix match since /apis matches everything (see the big comment on Director above)
+	isDiscoveryRequest := path == "/apis" || path == "/apis/"
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: move inside the `ws.RootPath() == "/apis"` case? we'll only have one webservice registered with that root path, right?

File: staging/src/k8s.io/apiserver/pkg/server/handler.go [hunk 12]
```diff
@@ -92,12 +114,57 @@ func (a *APIServerHandler) ListedPaths() []string {
...
+type director struct {
+	name               string
+	goRestfulContainer *restful.Container
+	nonGoRestfulMux    *mux.PathRecorderMux
+}
+
+func (d director) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	path := req.URL.Path
+
+	// if we are exactly /apis or /apis/, pass then we need special handling in loop.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/handler.go]: s/pass//

**Comment by liggitt** [comment]
a few questions/nits, LGTM otherwise

**Comment by deads2k** [comment]
comments addressed.  applying

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-node-e2e test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-unit test this
